src/libraries/System.Linq/src/System/Linq/Chunk.cs

```diff
@@ -52,24 +54,56 @@ private static IEnumerable<TSource[]> ChunkIterator<TSource>(IEnumerable<TSource
         {
             using IEnumerator<TSource> e = source.GetEnumerator();
 
+            // Before allocating anything, make sure there's at least one element.
             if (e.MoveNext())
             {
-                List<TSource> chunkBuilder = new();
-                int lastLength;
+                // Now that we know we have at least one item, allocate an initial storage array. This is not
+                // the array we'll yield.  It starts out small in order to avoid significantly overallocating
+                // when the source has many fewer elements than the chunk size.
+                var array = new TSource[Math.Min(size, 4)];
+                int i;
                 do
                 {
-                    do
+                    // Store the first item.
+                    array[0] = e.Current;
+                    i = 1;
+
+                    if (size == array.Length)
+                    {
+                        // For all but the first chunk, the array will already be correctly sized.
+                        // We can just store into it until either it's full or MoveNext returns false.
+                        TSource[] local = array; // avoid bounds checks from using a field-based array
```

Comment:
Can you clarify what you mean by that comment? `array` is also a local.

Reply:
> array is also a local.

Not in the IL. It becomes a field in the object in order to survive across yield points. I can clarify the comment.

Reply:
Ah right, almost forgot this is an iterator.